File: shibaura-institute-technology/exJava/go/Q33.go
package main

import "fmt"

func main() {
	pl := newPairList()
	pl.add("key1", 10)
	pl.add("key2", 20)
	pl.add("key3", 30)

	e1, e2 := pl.get(0)
	showAll(e1, e2)

	e1, e2 = pl.get(1)
	showAll(e1, e2)

	e1, e2 = pl.get(2)
	showAll(e1, e2)
}

type pairList struct {
	list1 []interface{}
	list2 []interface{}
}

func newPairList() *pairList {
	return &pairList{make([]interface{}, 0), make([]interface{}, 0)}
}
func (pl *pairList) add(e1 interface{}, e2 interface{}) {
	pl.list1 = append(pl.list1, e1)
	pl.list2 = append(pl.list2, e2)
}
func (pl *pairList) get(index int) (interface{}, interface{}) {
	return pl.list1[index], pl.list2[index]
}

// no need for holders, functions can return multiple values
func showAll(e1, e2 interface{}) {
	fmt.Printf("v1:v2 = %s:%d\n", e1, e2)
}

File: shibaura-institute-technology/exJava/java/Q33.java
import java.util.*;
import java.io.*;

public class Q33 {
  public static void main(String args[]) throws Exception {
    PairList<String, Integer> list1 = new PairList<String, Integer>();
    list1.add("key1", 10);
    list1.add("key2", 20);
    list1.add("key3", 30);

    Holder<String, Integer> holder = null;
    holder = list1.get(0);
    holder.showAll();

    holder = list1.get(1);
    holder.showAll();

    holder = list1.get(2);
    holder.showAll();
  }
}

class PairList<E1, E2> {
  List<E1> list1 = new ArrayList<E1>();
  List<E2> list2 = new ArrayList<E2>();

  public void add (E1 e1, E2 e2) {
    list1.add(e1);
    list2.add(e2);
  }

  public Holder<E1, E2> get (int index) {
    return new Holder(list1.get(index), list2.get(index));
  }
}

class Holder<V1, V2> {
  V1 v1;
  V2 v2;

  public Holder(V1 v1, V2 v2) {
    this.v1 = v1;
    this.v2 = v2;
  }

  public void showAll() {
    System.out.println("v1:v2 = " + v1 + ":" + v2);
  }
}

